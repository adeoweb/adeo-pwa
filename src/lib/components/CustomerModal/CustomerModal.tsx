import React, { Fragment, FunctionComponent, useState } from 'react';
import { Modal } from 'react-bootstrap';
import { useTranslation } from 'react-i18next';
import { useAppContext } from 'src/peregrine/lib/context/adeoweb/app';
import LoginForm from 'src/lib/components/Customer/LoginForm';
import LoginPageHeading from 'src/lib/components/Customer/pages/LoginPage/LoginPageHeading';
import { CustomerModalTypes } from 'src/lib/constants/customer';
import CloseIcon from 'src/lib/assets/icons/CloseIcon';
import ForgotPasswordModal from 'src/lib/components/ForgotPasswordModal';
import ForgotPasswordModalClasses from 'src/lib/components/ForgotPasswordModal/ForgotPasswordModal.scss';

interface ICustomerModalProps {
    modalType: CustomerModalTypes | null;
    handleClose: () => Promise<void>;
}

const CustomerModal: FunctionComponent<ICustomerModalProps> = ({
    modalType,
    handleClose
}) => {
    const { t } = useTranslation();
    const [, { setCustomerModal }] = useAppContext();

    // Keep modal type value in state to ensure proper render during close animation
    const [displayedModalType, setDisplayedModalType] = useState(modalType);

    if (modalType && modalType !== displayedModalType) {
        setDisplayedModalType(modalType);
    }

    let child: JSX.Element | null = null;

    const modalPropsBase = {
        centered: true,
        scrollable: false
    };

    let modalProps = {};

    switch (displayedModalType) {
        case CustomerModalTypes.LOG_IN:
            child = (
                <Fragment>
                    <LoginPageHeading
                        title={t('Login')}
                        paragraph={t(
                            'If you have an account with us, please log in.'
                        )}
                    />
                    <LoginForm
                        openForgotPassword={() => {
                            setCustomerModal(
                                CustomerModalTypes.PASSWORD_RECOVERY
                            );
                        }}
                        isSignedInCallback={handleClose}
                    />
                </Fragment>
            );
            break;
        case CustomerModalTypes.PASSWORD_RECOVERY:
            modalProps = {
                className: ForgotPasswordModalClasses.modal
            };
            child = <ForgotPasswordModal />;
            break;
    }

    return (
        <Modal
            show={Boolean(modalType)}
            onHide={handleClose}
            {...modalPropsBase}
            {...modalProps}
        >
            <button className="customer-modal-close-btn" onClick={handleClose}>
                <CloseIcon />
            </button>
            <div className="customer-modal">{child}</div>
        </Modal>
    );
};

export default CustomerModal;
