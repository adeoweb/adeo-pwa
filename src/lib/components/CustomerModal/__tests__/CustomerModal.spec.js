import React from 'react';
import testRenderer from 'react-test-renderer';
import { BrowserRouter as Router } from 'react-router-dom';
import CustomerModal from 'src/lib/components/CustomerModal';

jest.mock('src/peregrine/lib/talons/adeoweb/SignIn/useSignIn', () => {
    return {
        useSignIn: () => ({
            errors: [],
            handleSubmit: jest.fn(() => {}),
            isBusy: false
        })
    };
});

jest.mock('react-dom', () => ({
    createPortal: node => node
}));

jest.mock('react-bootstrap/Modal', () => ({ children = [] }) => (
    <div id="modal">{children}</div>
));

jest.mock('src/peregrine/lib/context/adeoweb/user', () => {
    const state = {
        currentUser: {
            firstname: 'firstname',
            lastname: 'lastname'
        },
        isSignedIn: true
    };
    const useUserContext = jest.fn(() => [state]);

    return { useUserContext };
});

jest.mock('src/peregrine/lib/context/adeoweb/app', () => {
    const api = {
        setCustomerModal: jest.fn(() => {})
    };
    const useAppContext = jest.fn(() => [null, api]);

    return { useAppContext };
});

test('Empty CustomerModal component renders correctly', () => {
    const tree = testRenderer.create(
        <Router>
            <CustomerModal modalType={null} handleClose={jest.fn(() => {})} />
        </Router>
    );

    expect(tree.toJSON()).toMatchSnapshot();
});

test('Non empty CustomerModal component renders correctly', () => {
    const tree = testRenderer.create(
        <Router>
            <CustomerModal
                modalType={'LOG_IN'}
                handleClose={jest.fn(() => {})}
            />
        </Router>
    );

    expect(tree.toJSON()).toMatchSnapshot();
});
