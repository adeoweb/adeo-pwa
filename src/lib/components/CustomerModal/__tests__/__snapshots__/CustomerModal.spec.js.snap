// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Empty CustomerModal component renders correctly 1`] = `
<div
  id="modal"
>
  <button
    className="customer-modal-close-btn"
    onClick={[MockFunction]}
  >
    <svg
      fill="none"
      height="25"
      viewBox="0 0 24 25"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clipRule="evenodd"
        d="M19 6.91L17.59 5.5L12 11.09L6.41 5.5L5 6.91L10.59 12.5L5 18.09L6.41 19.5L12 13.91L17.59 19.5L19 18.09L13.41 12.5L19 6.91Z"
        fill="#025198"
        fillRule="evenodd"
      />
    </svg>
  </button>
  <div
    className="customer-modal"
  />
</div>
`;

exports[`Non empty CustomerModal component renders correctly 1`] = `
<div
  id="modal"
>
  <button
    className="customer-modal-close-btn"
    onClick={[MockFunction]}
  >
    <svg
      fill="none"
      height="25"
      viewBox="0 0 24 25"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clipRule="evenodd"
        d="M19 6.91L17.59 5.5L12 11.09L6.41 5.5L5 6.91L10.59 12.5L5 18.09L6.41 19.5L12 13.91L17.59 19.5L19 18.09L13.41 12.5L19 6.91Z"
        fill="#025198"
        fillRule="evenodd"
      />
    </svg>
  </button>
  <div
    className="customer-modal"
  >
    <div
      className="heading"
    >
      <h2
        className="title"
      >
        Login
      </h2>
      <p>
        If you have an account with us, please log in.
      </p>
    </div>
    <form
      className=""
      noValidate={true}
      onSubmit={[Function]}
    >
      <label
        className="form-label"
      >
        customer:Email
      </label>
      <input
        className="form-control"
        name="email"
        onChange={[Function]}
        required={true}
        type="text"
        value=""
      />
      <div
        className="invalid-feedback"
      />
      <label
        className="form-label"
      >
        customer:Password
      </label>
      <input
        className="form-control"
        name="password"
        onChange={[Function]}
        required={true}
        type="password"
        value=""
      />
      <div
        className="invalid-feedback"
      />
      <div
        className="form-footer"
      >
        <button
          className="btn btn-primary"
          disabled={false}
          type="submit"
        >
          customer:Login
        </button>
        <button
          className="forget-pass btn btn-link"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          customer:Forgot your password?
        </button>
      </div>
    </form>
  </div>
</div>
`;
